pkg/agent/flags/flag_controller.go

```diff
@@ -0,0 +1,346 @@
+package flags
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// FlagController is responsible for retrieving flag values from the appropriate sources,
+// determining precedence, sanitizing flag values, and notifying observers of changes.
+type FlagController struct {
+	logger           log.Logger
+	defaultValues    *FlagValues
+	cmdLineValues    *FlagValues
+	storedFlagValues *storedFlagValues
+	sanitizer        *flagValueSanitizer
+	overrideMutex    sync.RWMutex
+	overrides        map[FlagKey]flagValueOverride
+	observers        map[FlagsChangeObserver][]FlagKey
+}
+
+func NewFlagController(logger log.Logger, defaultValues *FlagValues, cmdLineValues *FlagValues, storedFlagValues *storedFlagValues, sanitizer *flagValueSanitizer) *FlagController {
+	fc := &FlagController{
+		logger:           logger,
+		defaultValues:    defaultValues,
+		cmdLineValues:    cmdLineValues,
+		storedFlagValues: storedFlagValues,
+		sanitizer:        sanitizer,
+		overrides:        make(map[FlagKey]flagValueOverride),
+		observers:        make(map[FlagsChangeObserver][]FlagKey),
+	}
+
+	return fc
+}
+
+func get[T any](fc *FlagController, key FlagKey) T {
+	// If there is a temporary override, it takes precedence over anything else
+	overrideValue, ok := getOverrideValue[T](fc, key)
+	if ok {
+		return overrideValue
+	}
+
+	// See if we can find a stored value to use
+	storedValue, ok := getStoredValue[T](fc, key)
+	if ok {
+		return storedValue
+	}
+
+	// We were not able to find a suitable stored value, now cmd line flags take precedence
+	cmdLineValue, ok := getCmdLineValue[T](fc, key)
+	if ok {
+		return cmdLineValue
+	}
+
+	// No suitable cmd line flag provided, now fallback to the default values
+	return getDefaultValue[T](fc, key)
+}
+
+// getOverrideValue looks for an override value for the key and returns it, and true
+// if an override value is not found, a zero value and false will be returned
+func getOverrideValue[T any](fc *FlagController, key FlagKey) (T, bool) {
+	override, ok := fc.overrides[key]
+	if !ok {
+		return *new(T), false
+	}
+
+	var t T
+	var anyvalue any
+	switch reflect.TypeOf(t).Kind() {
+	case reflect.Bool:
+		boolValue, ok := override.Value().(bool)
+		if ok {
+			anyvalue = boolValue
+		}
+	case reflect.Int64:
+		int64Value, ok := override.Value().(int64)
+		if ok {
+			// Integers are sanitized to avoid unreasonable values
+			if fc.sanitizer != nil {
+				anyvalue = fc.sanitizer.Sanitize(key, int64Value)
```

Comment:
I opted for doing only on the `get` side because the values could still be modified after storing - i.e. modify the database and replace values that were previously sanitized. Although it could lead to `Sanitize` being called more frequently, it's a very cheap function so I'm not concerned about a performance hit, and overall this seemed like a simpler solution.

Reply:
That makes sense, thanks for explaining!